src/Microsoft.AspNet.Mvc.Core/TempDataDictionary.cs

```diff
@@ -0,0 +1,269 @@
+        public TempDataDictionary([NotNull] IHttpContextAccessor context, [NotNull] ITempDataProvider provider)
+        {
+            _provider = provider;
+            _loaded = false;
+            _contextAccessor = context;
+            _data = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
+        }
+
+        public int Count
+        {
+            get
+            {
+                Load();
+                return _data.Count;
+            }
+        }
+
+        public ICollection<string> Keys
+        {
+            get
+            {
+                Load();
+                return _data.Keys;
+            }
+        }
+
+        public ICollection<object> Values
+        {
+            get
+            {
+                Load();
+                return _data.Values;
+            }
+        }
+
+        bool ICollection<KeyValuePair<string, object>>.IsReadOnly
+        {
+            get { return ((ICollection<KeyValuePair<string, object>>)_data).IsReadOnly; }
+        }
+
+        public object this[string key]
+        {
+            get
+            {
+                Load();
+                object value;
+                if (TryGetValue(key, out value))
+                {
+                    _initialKeys.Remove(key);
+                    return value;
+                }
+                return null;
+            }
+            set
+            {
+                Load();
+                _data[key] = value;
+                _initialKeys.Add(key);
+            }
+        }
+
+        /// <inheritdoc />
+        public void Keep()
+        {
+            Load();
+            _retainedKeys.Clear();
+            _retainedKeys.UnionWith(_data.Keys);
+        }
+
+        /// <inheritdoc />
+        public void Keep(string key)
+        {
+            Load();
+            _retainedKeys.Add(key);
+        }
+
+        /// <inheritdoc />
+        public void Load()
+        {
+            if (_loaded)
+            {
+                return;
+            }
+
+            var providerDictionary = _provider.LoadTempData(_contextAccessor.HttpContext);
+            _data = (providerDictionary != null)
+                ? new Dictionary<string, object>(providerDictionary, StringComparer.OrdinalIgnoreCase)
+                : new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
+            _initialKeys = new HashSet<string>(_data.Keys, StringComparer.OrdinalIgnoreCase);
+            _retainedKeys.Clear();
+            _loaded = true;
+        }
+
+        /// <inheritdoc />
+        public void Save()
+        {
+            if (!_loaded)
+            {
+                return;
+            }
+
+            _data.RemoveFromDictionary((KeyValuePair<string, object> entry, TempDataDictionary tempData) =>
+            {
+                var key = entry.Key;
+                return !tempData._initialKeys.Contains(key)
+                    && !tempData._retainedKeys.Contains(key);
+            }, this);
```

Comment:
IMO just replace this with a loop rather than introducing our own vocabulary of collection methods :+1: